### ISSUE TITLE
RFC 5.x App Structure Changes

### ISSUE THREAD

**Issue by burzum** [cakephp/cakephp#12207]
Since it was decided to keep 4.x as BC as possible this is a RFC type ticket for 5.x:

This is a proposal to change the structure of the application and some of the framework folders as well based on https://github.com/php-pds/skeleton.

I'm actively using this structure already in an experimental hexagonal application architecture I'm developing for fun and learning purposes. The separation into for separate layers is really nice and could be used even if somebody is not going for a strict DDD apporach or hexagonal or onion architecture. I think the same structure can serve both kinds of people. If you want to go for a proper separation of the layers you'll use a lot interfaces and data transfer objects, if not you're free to ignore that concept.

* **Application**: Holds the console, http (web) and possibly other applications that use the same domain model. For example, if not php, you could have Android and Win32 folders here as well. The applications are the parts in this construct that will bring the other layers together.
* **(Domain)Model**: The domain model that should encapsulate the whole application logic and in an ideal case not be aware of any other layer. It's a representation of the business logic and rules of your application. However, no hard enforcement required of these concepts in Cake. If people want to put all kind of crap in here let them do it but provide people who care a good structure.
* **Infrastructure**: Is everything that deals other APIs like databases, RESTful resources, sending SMS, Email and so on.
* **Presentation**: Should contain only logic required for rendering.

I've moved the templates into the `resources` folder because that's what they're in my opinion they're not really code, especially when going for twig. The resources folder will hold other resources as well that are used to build the application by a tool like Webpack.

I'm aware of the discussion to provide "packs" for different frontend libs like Vue, React or Angular. If we don't want to rely on the ever mutating and fast moving JS mess, we could use something like https://github.com/consolidation/Robo instead to provide a build process out of the box. But I doubt it's going to work as good as having different packs that can be installed and used via yarn. However, point is to separate the resources from the app.

The core should reflect the folder structure as well.

Here is the output of my current (work in progress) application structure I'm using for my hexagonal architecture application:

```sh
├───bin
├───config
│   └───migrations
├───docs
├───logs
├───public
│   ├───css
│   ├───fonts
│   ├───img
│   └───js
├───resources
│   ├───locales
│   ├───files
│   ├───js
│   ├───sass
│   ├───scss
│   └───templates
│       ├───Articles
│       ├───Element
│       ├───Error
│       ├───Layout
│       └───Pages
├───src
│   ├───Application
│   │   ├───Console
│   │   │   ├───Dispatcher
│   │   │   └───Shell
│   │   ├───Http
│   │   │   ├───Command
│   │   │   │   ├───Articles
│   │   │   │   └───Pages
│   │   │   ├───CommandHandler
│   │   │   │   ├───Articles
│   │   │   │   └───Pages
│   │   │   ├───Controller
│   │   │   │   ├───Pages
│   │   │   │   └───Users
│   │   │   ├───Dispatcher
│   │   │   ├───HandlerResult
│   │   │   └───Service
│   │   └───Service
│   ├───Model
│   │   ├───Entity
│   │   ├───Repository
│   │   └───Service
│   ├───Infrastructure
│   │   ├───Container
│   │   ├───Email
│   │   ├───Persistence
│   │   │   ├───Database
│   │   │   │   └───Table
│   │   │   ├───ElasticSearch
│   │   │   └───Session
│   │   └───Service
│   └───Presentation
│       └───Renderer
├───tests
│   ├───Acceptance
│   └───Unit
│       └───TestCase
├───tmp
│   └───cache
│       ├───models
│       └───persistent
└───vendor
```

**Comment by markstory** [comment]
Do you think that CakePHP is well suited for these architectural styles?

**Comment by burzum** [comment]
@markstory with enough interfaces and well separated packages and a DI container a clear yes. Because *if* somebody wants to go for a strict domain model your domain model must stay agnostic to the other layers. But this folder structure is IMHO in any case more clear than what we currently have. You don't have to follow it or use it as intended. People already ignore what we have. Smart people will think about why it is like this. We can add a paragraph for each layer to the docs.

**Comment by dereuromark** [comment]
As per [this](https://github.com/cakephp/cakephp/pull/11931) templates should probably stay primary level (no need to change again), and you also forgot /locale as new primary level folder for former src/Locale po files etc.

I think some of the proposed changes might go a bit far, if we settled somewhere along the middle for a cleaner separation between communication and business code, that would already be a big win.

In general this discussion might be a bit soon when 4.x is still in its superearly shoes still yet.

**Comment by burzum** [comment]
@dereuromark I disagree with the change to move these things into the top level. Both should go into resource. But good point, I've added the locales to the tree above.

Who goes this to far? Actually we haven't really decided how to improve the model layer. The `Model` folder could stay. The other stuff goes mostly in new folders.

Also it didn't took me much time to re-arrange the folders. This is an easy change. And I personally want to see that Cake5 drops old stuff and architecture and goes for "enterprise" user cases. Simply because I think this is the audience then would provide proper feedback and might also invest money in the framework in form of donations.

**Comment by markstory** [comment]
I worry about trying to design for an enterprise audience. Doing so can easily alienate other audiences due to preceived/actual complexity. Historically, CakePHP has not been optimized for enterprise architectures and it would be a big shift in messaging and positioning to change course. 

It sounds like we would have a bunch of work to do on cake itself before this application structure would be useful. Perhaps we should start with some of those ideas first?

**Comment by burzum** [comment]
I expected an answer like that so if at all this would happen in Cake 6 or 7. Which sadly increases my willingness to switch to another framework at work. I doubt well start the CRM this year so we still have some time but already started to evaluate Laravel and will check Zend as well again after years not looking at it.

But again, I've written this now multiple times, I don't see this structure as a blocker for other people who disagree or don't know anything about architecture.

I don't think there is much effort in changing the structure. It didn't took me a half day to switch the structure I think and I didn't know always where exactly to change something, so I had to figure it out.

@markstory what is the audience *you* want to support? The learn resistant and copy and paste audience that doesn't read docs is nothing I care about nor do I think they are a benefit for the framework. I don't see how the folder structure change will add complexity. It will be pretty much the same stuff just in other folders. If that's to hard to learn and switch for some people, well... honestly, they should find another job. The way the MVC part works wouldn't change at all.

**Comment by markstory** [comment]
>  But I begin to have doubts that Cake will make the kind of improvements we're looking for any time soon.

I'm not clear on what the improvements you're hoping to see is. Do we have open issues for the gaps you see?

> I don't see how the folder structure change will add complexity. Decoupling the parts Cake is made of won't hurt them either 

Wide and deep folder structures are indicative of complexity. More folders is roughly equal to more top-level concepts that need to be learned and composed together. As an example compare the layout of an emberjs application and a backbone.js application. Ember has significantly more folders, and concepts. It is also significantly more complex.

> Decoupling the parts Cake is made of won't hurt them either but it will make the framework better decoupled and more modular. 

I'm not against decreasing coupling and increasing the modularity of CakePHP. We've made significant progress against that in the past, and we have plans to make more subsystems available as standalone packages.

> what is the audience you want to support? The learn resistant and copy and paste audience that doesn't read docs is nothing I care about nor do I think they are a benefit for the framework.

Historically our audience has included very new developers and more seasoned folks who are looking for a quick to learn, simple to get started framework. This pairs well with the audience that PHP draws. To be honest I would be surprised if there are many large-scale enterprise applications being born into PHP. Instead I would expect folks starting those projects to choose languages that lend themselves better to large applications (go, scala, java) because of their type systems and tooling.

Instead of shifting focus towards the smaller pool of big architecture applications, I would prefer to explore ways we can continue to meet the needs of the small/simple application and be able to scale up to the larger application use case.

> If that's to hard to learn and switch for some people, well... honestly, they should find another job.

This is a poor position to take and makes you sound like you're unable to remember what it is like to learn something new.

**Comment by burzum** [comment]
> I'm not clear on what the improvements you're hoping to see is. Do we have open issues for the gaps you see?

https://github.com/cakephp/cakephp/issues/12114 also I looked briefly at some of our packages, it looks like we have sometimes dependencies declared that aren't used and sometimes we could decouple other packages. There is probably more but I haven't kept a record of it.

> Wide and deep folder structures are indicative of complexity. More folders is roughly equal to more top-level concepts that need to be learned and composed together.

We have a ton of top level folders? But I don't think they group the responsibility very well. However, I think this is totally opinionated and a personal preference. A smart approach is usually to ask why something is like it is. Smart developers will question the folder structure, no matter if flat or deep to understand the concept behind it. I still don't think the change would hurt. People who don't care will the best perceive it as a nasty additional click on the folder tree and don't ask questions about the reason.

> Historically our audience has included very new developers and more seasoned folks who are looking for a quick to learn, simple to get started framework. This pairs well with the audience that PHP draws. To be honest I would be surprised if there are many large-scale enterprise applications being born into PHP. 

I think most enterprise apps are not making public what they use, no matter if php or Java. And yes php attracts more inexperienced people because php is easy. But php has as well a bad reputation as a language. In some aspects this is right but php has evolved as well. php allows inexperienced people to get results, but honestly, the results are just a security and maintenance disaster. I dislike the idea of orienting the framework or code towards this audience instead of looking up.

> Instead I would expect folks starting those projects to choose languages that lend themselves better to large applications (go, scala, java) because of their type systems and tooling.

This is actually something I'm looking forward to do, but I would love to work with c#. I want to go for c# development in the future, shifting away from php.

> This is a poor position to take and makes you sound like you're unable to remember what it is like to learn something new.

I learn every week something new, no need to remember. But if you struggle to adept to a simple change of a folder structure, there is a much much more fundamental issue than learning programming or a framework. If you want to drag these people along OK, I don't want to do that. If the framework is built for educational purpose, OK, I'm all with your argumentation. If not I would assume it's made for business and I would expect even a junior developer to be capable of dealing with a simple folder change.

And don't get me wrong, I really like to teach people and transfer knowledge and I do so, but there are people who are unwilling or unable to change their attitude towards "learning how to learn".

**Comment by burzum** [comment]
@markstory another example is the routing. We can't provide the router as standalone package yet, or just hard. The dependency on `\Cake\Core\Configure` could be resolved by injection the config (remember my request for configuration objects? ~1.5 years ago?) instead of reading it from inside the router code. The `RequestActionTrait` should not go into the router package but more likely into 
`\Cake\Mvc\Http` or `\Cake\Http`. And overall, but I don't think it will be easy, `\Cake\Http\ServerRequest` should be changed to `\Psr\Http\ServerRequestInterface` or whatever the exact namespace is.

**Comment by robertpustulka** [comment]
I think that folder/app structure is not really something that the framework should enforce. But I'm all :+1: for decoupling parts of the framework so it would be easier to use or replace parts of the application if someone wants to leverage hexagional design for more complex/enterprise projects. Just look at the modern PHP world. Frameworks become slimmer, lighter. Some devs try framework-less approach. If we had more code released as stand-alone, more interop packages, more people could use that in their project and therefore we might attract more audience.

**Comment by ravage84** [comment]
> Since it was decided to keep 4.x as BC as possible this is a RFC type ticket for 5.x:

Well the idea was to keep major upgrades in general as BC as possible to prevent upgrades like CakePHP 2.x to 3.0 in the future. So a change like proposed by you, @burzum, would be contrary of this idea. We would need to introduce such a change within the lifespan of a major, either 4.x or 5.x.

> I've moved the templates into the resources folder because that's what they're in my opinion they're not really code, especially when going for twig. 

I disagree. Even when stricltly using a templating engine, the template files are still code. I would put unsorted files like XSDs, CSVs etc. in the ressources folder. See https://github.com/php-pds/skeleton/issues/12.

The same goes for the locales, imho.

> As per this templates should probably stay primary level (no need to change again), and you also forgot /locale as new primary level folder for former src/Locale po files etc.

Sounds more reasonable to me. :+1: 

I'm OK with ``public``, even though I think it's as bad as many other names. See my disagreement on https://github.com/php-pds/skeleton/issues/10.

As for the source folder. At the very first glance I'm very much reminded on my IT school days when we had to learn Java. I don't think this is a good thing. _I don't see "Cake" in there_ but more like the word "Enteprise".

> I worry about trying to design for an enterprise audience. Doing so can easily alienate other audiences due to preceived/actual complexity. Historically, CakePHP has not been optimized for enterprise architectures and it would be a big shift in messaging and positioning to change course.

Amen. :pray: 

> It sounds like we would have a bunch of work to do on cake itself before this application structure would be useful. Perhaps we should start with some of those ideas first?

I would very much prefer this, instead of doing "everything" differently at once. This gives people the possiblity and time to learn and adapt to the changes.

````
├───src
│   ├───Application
│   │   ├───Console
│   │   │   ├───Dispatcher
│   │   │   └───Shell
│   │   ├───Http
│   │   │   ├───Command
│   │   │   │   ├───Articles
│   │   │   │   └───Pages
│   │   │   ├───CommandHandler
│   │   │   │   ├───Articles
│   │   │   │   └───Pages
│   │   │   ├───Controller
│   │   │   │   ├───Pages
│   │   │   │   └───Users
│   │   │   ├───Dispatcher
│   │   │   ├───HandlerResult
│   │   │   └───Service
│   │   └───Service
│   ├───Model
│   │   ├───Entity
│   │   ├───Repository
│   │   └───Service
│   ├───Infrastructure
│   │   ├───Container
│   │   ├───Email
│   │   ├───Persistence
│   │   │   ├───Database
│   │   │   │   └───Table
│   │   │   ├───ElasticSearch
│   │   │   └───Session
│   │   └───Service
│   └───Presentation
│       └───Renderer
``````

I count four ``Service`` folders. Going from none to four is quite a jump.

Also, for example, if I'm working on the Articles command, I have to go five folders deep, /src/Application/Http/Command/ -> Articles. Again, I'm reminded of Java.

> I think some of the proposed changes might go a bit far, if we settled somewhere along the middle for a cleaner separation between communication and business code, that would already be a big win.

Totally agree.

> @markstory what is the audience you want to support?

I'd love we would talk about it and settle for a generally and officially targeted audience. But targeting one audience does not totally exclude users from other audiences.

> But we've already lost on this side if you look how many other libs and frameworks use parts of the Laravel or Symfony framework because it was a decade ago possible to just use part of it. 

While I agree on the fact, I do not agree on the reason. Other frameworks started years ago to split their framework into separate packages which could be included through Composer. It took CakePHP until 3.0 to do that.

> Wide and deep folder structures are indicative of complexity. More folders is roughly equal to more top-level concepts that need to be learned and composed together.

Agree.

> Instead of shifting focus towards the smaller pool of big architecture applications, I would prefer to explore ways we can continue to meet the needs of the small/simple application and be able to scale up to the larger application use case.

I, too, would welcome such an approach. Documentation, recommendations, dedicated learning resources etc. could be possible ways to tackle that.

> Take a look at the Zend framework. I've just installed it recently and it offers you a lot choices for the router, the view and the error handler. 

I think this is not a fair comparison. The Zend framework has always been a very modular, "it is your choice how you stitch it together"- framework while CakePHP has always been a very opinionated fully prepackaged general purpose web framework.
Nonetheless I, too, think we can do better at being more modular.

> I hope this is perceived as real criticism and not just bitching around. smiley

I, for one, perceive it as such and I support you in your general attempt, but at the same time I don't think your approach is the correct one.

> I think that folder/app structure is not really something that the framework should enforce.

@robertpustulka  Well one of CakePHP's principles is "Convention over Configuration", so at least in some regard this already disagrees with you.  But at the same time, a framework should still be flexible enough for change and adaptation.

> Some devs try framework-less approach. If we had more code released as stand-alone, more interop packages, more people could use that in their project and therefore we might attract more audience.

True!

All in all, @burzum thank you very much for your efforts to push CakePHP. :heart:

**Comment by robertpustulka** [comment]
@ravage84 I meant that CakePHP shouldn't enforce DDD-style folder structure in this case.

**Comment by burzum** [comment]
I'm OK with @dereuromark to close this as this isn't going to go anywhere soon and we'll very likely drop CakePHP anyway for new projects (except the ORM), probably in favor of a framework less approach or something totally different.

**Comment by robertpustulka** [comment]
> we'll very likely drop CakePHP anyway for new projects (except the ORM), probably in favor of a framework less approach or something totally different.

This is what I'm talking about. If you start your next big (enterprise) project you may not want to use CakePHP (as this is more like a RAD framework) but you might use it's great components. ORM is one of those. I think we could improve stand-alone packages instead and target enterprise (or whatever you call it) audience with that.

**Comment by dakota** [comment]
Is there any reason why somebody can't create an alternative "enterprisy" app skeleton? With enough traction, we could even consider making it official.

**Comment by burzum** [comment]
@dakota I've tried but it's not as easy because a ton of things are very interwoven in the framework. :(

As soon as you change something that requires a locator you need a new locator. Sometimes you can't replace one very easy. The locators work nice as long as you don't want to break things up or modify and I never needed a DI container because until recently I was strictly sticking to the CakePHP world and working inside the boundaries of the framework. But as soon as you leave the comfort zone things aren't that easy anymore.

**Comment by lorenzo** [comment]
@burzum can you list some concrete examples? I'm curious about your experiments

**Comment by burzum** [comment]
@lorenzo see the ticket about the Mailer, the configuration being tied to the Configure. Everything that uses event is hard coupled to the cakephp events. The logging is not PSR compatible. I had to write my own controller and console dispatcher after I've change the structure. Also the dispatching is a little to much coupled as far as I remember. See my ticket about refactoring the email code as well.

I had to override BaseApplication::getDispatcher() instead of injection a dispatcher. Also why is the BaseApplication using ConsoleApplicationInterface as well? It's never needed in a web context. I think we should have HttpApplication and ConsoleApplication extending BaseApplication and implementing the interfaces instead of having BaseApplication implementing both interfaces.
